algorithms/linfa-bayes/src/bernoulli_nb.rs

```diff
@@ -0,0 +1,336 @@
+        let feature_count = if count_old > &0 {
+            feature_count_old + feature_count_new
+        } else {
+            feature_count_new
+        };
+        // Apply smoothing to feature counts
+        let feature_count_smoothed = feature_count.clone() + self.alpha();
+        // Compute total count (smoothed)
+        let count = F::cast(x_new.nrows()) + self.alpha() * F::cast(2);
+        // Compute log probabilities of each feature
+        let feature_log_prob = feature_count_smoothed.mapv(|x| x.ln() - F::cast(count).ln());
+        (feature_log_prob.to_owned(), feature_count.to_owned())
+    }
+}
+
+/// Fitted Bernoulli Naive Bayes classifier.
+///
+/// See [BernoulliNbParams] for more information on the hyper-parameters.
+///
+/// # Model assumptions
+///
+/// The family of Naive Bayes classifiers assume independence between variables. They do not model
+/// moments between variables and lack therefore in modelling capability. The advantage is a linear
+/// fitting time with maximum-likelihood training in a closed form.
+///
+/// # Model usage example
+///
+/// The example below creates a set of hyperparameters, and then uses it to fit
+/// a Bernoulli Naive Bayes classifier on provided data.
+///
+/// ```rust
+/// use linfa_bayes::{BernoulliNbParams, BernoulliNbValidParams, Result};
+/// use linfa::prelude::*;
+/// use ndarray::array;
+///
+/// let x = array![
+///     [-2., -1.],
+///     [-1., -1.],
+///     [-1., -2.],
+///     [1., 1.],
+///     [1., 2.],
+///     [2., 1.]
+/// ];
+/// let y = array![1, 1, 1, 2, 2, 2];
+/// let ds = DatasetView::new(x.view(), y.view());
+///
+/// // create a new parameter set with smoothing parameter equals `1`
+/// let unchecked_params = BernoulliNbParams::new()
+///     .alpha(1.0);
+///
+/// // fit model with unchecked parameter set
+/// let model = unchecked_params.fit(&ds)?;
+///
+/// // transform into a verified parameter set
+/// let checked_params = unchecked_params.check()?;
+///
+/// // update model with the verified parameters, this only returns
+/// // errors originating from the fitting process
+/// let model = checked_params.fit_with(Some(model), &ds)?;
+/// # Result::Ok(())
+/// ```
+#[derive(Debug, Clone, PartialEq)]
+pub struct BernoulliNb<F: PartialEq, L: Eq + Hash> {
+    class_info: HashMap<L, MultinomialClassInfo<F>>,
+    binarize: Option<F>,
+}
+
+impl<F: Float, L: Label> BernoulliNb<F, L> {
+    /// Construct a new set of hyperparameters
+    pub fn params() -> BernoulliNbParams<F, L> {
+        BernoulliNbParams::new()
+    }
+
+    // Binarize data if the threshold is set
+    fn binarize<'a, D>(&'a self, x: &'a ArrayBase<D, Ix2>) -> ArrayBase<OwnedRepr<F>, Ix2>
+    where
+        D: Data<Elem = F>,
+    {
+        let thr = self.binarize.unwrap_or(F::zero());
+        let xbin = x.map(|v| if v > &thr { F::one() } else { F::zero() });
+        return if self.binarize.is_some() {
+            xbin.to_owned()
+        } else {
+            x.to_owned()
+        };
```

Comment:
The signature should return `CowArray2<F>` so that you don't create a new array when there's no threshold.

The body is better written as:
```rust
if let Some(thr) = self.binarize {
  x.map(|v| if v > &thr { F::one() } else { F::zero() }).into()
} else {
  x.into()
}
```